framework/src/service/ServiceReferenceBasePrivate.cpp

```diff
@@ -103,64 +114,58 @@ InterfaceMapConstPtr ServiceReferenceBasePrivate::GetPrototypeService(const Bund
 
 std::shared_ptr<void> ServiceReferenceBasePrivate::GetService(BundlePrivate* bundle)
 {
-  auto s = ExtractInterface(GetServiceInterfaceMap(bundle), interfaceId);
-  if (!s)
-  {
-    registration->Lock(), --registration->dependents[bundle];
-  }
-  return s;
+  return ExtractInterface(GetServiceInterfaceMap(bundle), interfaceId);
 }
 
 InterfaceMapConstPtr ServiceReferenceBasePrivate::GetServiceInterfaceMap(BundlePrivate* bundle)
 {
   InterfaceMapConstPtr s;
   if (!registration->available) return s;
-
   std::shared_ptr<ServiceFactory> serviceFactory;
-  int count = 0;
 
   {
     auto l = registration->Lock(); US_UNUSED(l);
     if (!registration->available) return s;
     serviceFactory = std::static_pointer_cast<ServiceFactory>(
           registration->GetService_unlocked("org.cppmicroservices.factory"));
-    count = registration->dependents[bundle];
-  }
 
-  if (!serviceFactory)
-  {
-    auto l = registration->Lock(); US_UNUSED(l);
-          s = registration->service;
-    if (s && !s->empty()) ++registration->dependents[bundle];
-  }
-  else
-  {
-    if (count == 0)
-    {
-      s = GetServiceFromFactory(bundle, serviceFactory);
+    registration->dependents.insert(std::make_pair(bundle, 0));
 
-      auto l = registration->Lock(); US_UNUSED(l);
-      if (registration->dependents[bundle] == 0)
-      {
-        registration->bundleServiceInstance.insert(std::make_pair(bundle, s));
-      }
-      else
+    if (!serviceFactory)
+    {
+      s = registration->service;
+      if (s && !s->empty())
       {
-        // There was a race and we now have one instance too much. Return the
-        // already produced instance and ignore the additional one.
-        s = registration->bundleServiceInstance[bundle];
+        ++registration->dependents.at(bundle);
       }
-
-      if (s && !s->empty()) ++registration->dependents[bundle];
+      return s;
     }
-    else
+
+    if (registration->bundleServiceInstance.end() != registration->bundleServiceInstance.find(bundle))
     {
-      auto l = registration->Lock(); US_UNUSED(l);
-      // return the already produced instance
-      s = registration->bundleServiceInstance[bundle];
-      if (s && !s->empty()) ++registration->dependents[bundle];
+      s = registration->bundleServiceInstance.at(bundle);
+      ++registration->dependents.at(bundle);
+      return s;
     }
   }
+
+  // Calling into a service factory could cause re-entrancy into the
+  // framework and even, theoretically, into this function. Ensuring
+  // we don't hold a lock while calling into the service factory eliminates
+  // the possibility of a deadlock. It does not however eliminate the
+  // possbilituy of infinite recursion.
+  s = GetServiceFromFactory(bundle, serviceFactory);
+
+  auto l = registration->Lock(); US_UNUSED(l);
+
+  registration->dependents.insert(std::make_pair(bundle, 0));
+
+  if (s && !s->empty())
```

Comment:
I agree that we cannot control the order, but we do not have to. The only case I think we should cover is that if we get a null service but have a non-null service cached because of a race, return the non-null service (just because we can). A service factory is allowed to return null and there is nothing we can do against that.